src/main/java/org/sonatype/nexus/ci/nxrm/v3/CreateTagBuilder.java

```diff
@@ -0,0 +1,229 @@
+
+import static com.sonatype.nexus.api.common.NexusStringUtils.isBlank;
+import static com.sonatype.nexus.api.common.NexusStringUtils.isNotBlank;
+import static hudson.Util.fixEmptyAndTrim;
+import static hudson.model.Result.FAILURE;
+import static hudson.util.FormValidation.error;
+import static hudson.util.FormValidation.ok;
+import static org.sonatype.nexus.ci.config.NxrmVersion.NEXUS_3;
+import static org.sonatype.nexus.ci.nxrm.Messages.CreateTag_DisplayName;
+import static org.sonatype.nexus.ci.nxrm.Messages.CreateTag_Error_TagAttributesJson;
+import static org.sonatype.nexus.ci.nxrm.Messages.CreateTag_Error_TagAttributesPath;
+import static org.sonatype.nexus.ci.nxrm.Messages.CreateTag_Validation_TagAttributesJson;
+import static org.sonatype.nexus.ci.nxrm.Messages.CreateTag_Validation_TagNameEmpty;
+import static org.sonatype.nexus.ci.nxrm.Messages.CreateTag_Validation_TagNameInvalid;
+import static org.sonatype.nexus.ci.util.RepositoryManagerClientUtil.nexus3Client;
+
+public class CreateTagBuilder
+    extends Builder
+    implements SimpleBuildStep
+{
+  static final Gson GSON = new Gson();
+
+  static final Type ATTRIBUTE_TYPE = new TypeToken<Map<String, Object>>() { }.getType();
+
+  private final String nexusInstanceId;
+
+  private final String tagName;
+
+  @CheckForNull
+  private String tagAttributesPath;
+
+  @CheckForNull
+  private String tagAttributesJson;
+
+  @DataBoundConstructor
+  public CreateTagBuilder(final String nexusInstanceId, final String tagName)
+  {
+    this.nexusInstanceId = nexusInstanceId;
+    this.tagName = tagName;
+  }
+
+  public String getNexusInstanceId() {
+    return nexusInstanceId;
+  }
+
+  public String getTagName() {
+    return tagName;
+  }
+
+  @CheckForNull
+  public String getTagAttributesPath() {
+    return tagAttributesPath;
+  }
+
+  @DataBoundSetter
+  public void setTagAttributesPath(@CheckForNull final String tagAttributesPath) {
+    this.tagAttributesPath = fixEmptyAndTrim(tagAttributesPath);
+  }
+
+  @CheckForNull
+  public String getTagAttributesJson() {
+    return tagAttributesJson;
+  }
+
+  @DataBoundSetter
+  public void setTagAttributesJson(@CheckForNull final String tagAttributesJson) {
+    this.tagAttributesJson = fixEmptyAndTrim(tagAttributesJson);
+  }
+
+  @Override
+  public void perform(@Nonnull final Run run, @Nonnull final FilePath workspace, @Nonnull final Launcher launcher,
+                      @Nonnull final TaskListener listener) throws InterruptedException, IOException
+  {
+    RepositoryManagerV3Client client = null;
+    Map<String, Object> tagAttributes = (isNotBlank(tagAttributesPath) ||
+        isNotBlank(tagAttributesJson)) ? new HashMap<>() : null;
+    EnvVars env = run.getEnvironment(listener);
+
+    try {
+      client = nexus3Client(nexusInstanceId);
+    }
+    catch (RepositoryManagerException e) {
+      e.getResponseMessage();
+      failBuildAndThrow(run, listener, e.getResponseMessage().orElse(e.getMessage()), new IOException(e));
```

Comment:
Trivial: Is this `e.getResponseMessage()` a leftover from previous iterations?

Reply:
Or just some debug code that I forgot to remove; thanks for pointing it out